configtelemetry/telemetry.go

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package configtelemetry
+
+import (
+	"errors"
+	"fmt"
+	"go/ast"
+	"go/types"
+	"strings"
+
+	"golang.org/x/tools/go/analysis"
+)
+
+var (
+	telemetryPkgPath = "github.com/mattermost/mattermost-server/v5/services/telemetry"
+	modelPkgPath     = "github.com/mattermost/mattermost-server/v5/model"
+)
+
+// Analyzer reports config fields where telemetry is not sent
+var Analyzer = &analysis.Analyzer{
+	Name: "configtelemetry",
+	Doc:  "reports config fields where telemetry is not sent (unless it is specifically indicated  that no telemetry will be sent for that field)",
+	Run:  run,
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	if pass.Pkg.Path() != telemetryPkgPath {
+		return nil, nil
+	}
+
+	// we need to find model.Config definition to parse its Type Spec
+	var configObject *types.TypeName
+	for _, obj := range pass.TypesInfo.Uses {
+		if tn, ok := obj.(*types.TypeName); ok && obj.Name() == "Config" && obj.Pkg().Path() == modelPkgPath {
+			configObject = tn
+			break
+		}
+	}
+
+	if configObject == nil {
+		return nil, errors.New("could not find model.Config type")
+	}
+
+	configMap, err := typeFieldMap(pass.Fset, configObject)
+	if err != nil {
+		return nil, fmt.Errorf("could generate fields map: %w", err)
+	}
+
+	var allExpressions []string
+
+	// We look every expression in the file so that we can
+	// check if the expected references exist in this package.
+
+	// NOTE: this approach does not check if the reference (config field) is added
+	// to telemetry queue, it assumes that if a config field is used in the file,
+	// it probably added to the telemetry. This is a naive approach but should work.
+	// Also, we can narrow down this for specific function calls, declarations etc.
```

Comment:
I am wondering how hard would it be to walk up a node and verify that the expression is used as a map value? Or even verify that all such expressions are only used inside `ts.sendTelemetry` method.

Just checking for these expressions within a package seems very broad to me. Would be good to have something more robust if possible.

Reply:
There are some occasions where config values are not directly used in `ts.sendTelemetry` (e.g. `PluginSettings`) so it gets quickly complicated to track those. I think checking if the expression is used as a map value won't be that hard.

Even we check if we the values are used as map value or in `sendTelemetry` method arg, it's still possible to trick the checker. At some point we need to trust the developer. I can add the map-value layer if you want. I'm 0/5 on either approach.

Reply:
Gave it some more thought. I agree with you. However, let's add an escape hatch as a custom comment to ignore a config field to let a developer use it without letting the vet pass think of it as something to be used in telemetry.